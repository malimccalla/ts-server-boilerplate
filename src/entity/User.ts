import * as bcrypt from 'bcryptjs';
import {
  BaseEntity,
  BeforeInsert,
  Column,
  Entity,
  PrimaryGeneratedColumn,
} from 'typeorm';

@Entity('users')
export class User extends BaseEntity {
  @PrimaryGeneratedColumn('uuid') id: string;

  @Column('varchar', { length: 255 })
  email: string;

  @Column('text') password: string;

  @Column('boolean', { default: false })
  confirmed: boolean;

  @Column('boolean', { default: false })
  locked: boolean;

  @BeforeInsert()
  async hashPassword() {
    const password = await bcrypt.hash(this.password, 10);
    this.password = password;
  }
}
