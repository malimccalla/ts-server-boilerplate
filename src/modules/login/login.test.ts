import * as faker from 'faker';
import { request } from 'graphql-request';
import { Connection } from 'typeorm';

import { loginMutation, registerMutation } from '../../test/ast';

import { User } from '../../entity/User';
import { createTestConn } from '../../test/createTestConn';

const host = process.env.TEST_HOST as string;

describe('Login', () => {
  let conn: Connection;

  beforeEach(async () => {
    conn = await createTestConn();
  });

  afterEach(async () => {
    await conn.close();
  });

  test('should login a user', async () => {
    const email = faker.internet.email();
    const password = faker.internet.password(10);

    const register = registerMutation(email, password);
    await request(host, register);

    // Manually confirm the users email address
    await User.update({ email }, { confirmed: true });

    const login = loginMutation(email, password);
    const res = (await request(host, login)) as {
      login: GQL.ILoginResponse;
    };

    expect(res.login.ok).toBe(true);
    expect(res.login.user!.email).toEqual(email);
  });

  test('should not login a user without email confirmation', async () => {
    const email = faker.internet.email();
    const password = faker.internet.password(10);

    const register = registerMutation(email, password);
    await request(host, register);

    const login = loginMutation(email, password);
    const res = (await request(host, login)) as {
      login: GQL.ILoginResponse;
    };

    expect(res.login.ok).toBe(false);
    expect(res.login).toMatchSnapshot();
  });
});
