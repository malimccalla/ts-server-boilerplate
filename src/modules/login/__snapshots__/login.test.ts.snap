// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should not login a user without email confirmation 1`] = `
Object {
  "errors": Array [
    Object {
      "message": "Please confirm your email address",
      "path": "email",
    },
  ],
  "ok": false,
  "user": null,
}
`;
